text/0000-promise-based-initializers.md

```diff
@@ -0,0 +1,86 @@
+- Start Date: 2020-01-09
+- Relevant Team(s): Ember.js
+- RFC PR: https://github.com/emberjs/rfcs/pull/572
+- Tracking: (leave this empty)
+
+# Promise-based Initializers
+
+## Summary
+
+This RFC proposes the ability to return a promise from an initializer or instance-initializer
+and pause application boot until that promise resolves.
+
+## Motivation
+
+Currently, the earliest place that developers can write asynchronous code is in the ApplicationRoute.
+For some use cases, this is not early enough, and it is useful to be able to run asynchronous code
+before the application begins routing.
+
+For example, when booting an Ember application in an embedded environment, it may be necessary to
+listen for events from the environment to determine the initial state to determine the initial URL.
```

Comment:
would using the beforeModel in the application route work?

Reply:
For this particular use cases, the application route is too late because there are subtle differences between what it means to *start* routing at a URL and what it means to redirect to a URL. When an Ember app runs in a browser, it does the former.

That said, I think promise based initializers will have benefits beyond just determining the initial state of the application.

Reply:
You can't redirect from the application route though. It,s the parent of everything.

Reply:
redirecting is just a fancy word for saying " cancel the current transition and start a new one". You *can* intercept a transition in the `ApplicationRoute` that would have ended at `/foo` and and redirect to `/bar`, for example.